packages/visx-shape/src/util/getSplitLineSegments.ts

```diff
@@ -1,80 +1,110 @@
 
-    const totalPathLength = pathElement.getTotalLength();
-    const totalPieces = segments.reduce((sum, curr) => sum + curr.length, 0);
-    const pieceSize = totalPathLength / totalPieces;
+    const numSegments = pointsInSegments.length;
+    const lineSegments: LineSegments = pointsInSegments.map(() => []);
 
-    let cumulativeSize = 0;
+    if (segmentation === 'x' || segmentation === 'y') {
+      const segmentBegins = pointsInSegments.map(
+        points => points.find(p => typeof p[segmentation] === 'number')?.[segmentation],
+      );
 
-    const lineSegments = segments.map(segment => {
-      const segmentPointCount = segment.length;
-      const coords: { x: number; y: number }[] = [];
+      const first = pathElement.getPointAtLength(0);
+      const last = pathElement.getPointAtLength(totalLength);
+      const isIncreasing = last[segmentation] > first[segmentation];
+      const isBeyondSegments = isIncreasing
+        ? segmentBegins.map(begin =>
+            typeof begin === 'undefined' ? TRUE : (xOrY: number) => xOrY >= begin,
+          )
+        : segmentBegins.map(begin =>
+            typeof begin === 'undefined' ? TRUE : (xOrY: number) => xOrY <= begin,
+          );
 
-      for (let i = 0; i < segmentPointCount + sampleRate; i += sampleRate) {
-        const distance = (cumulativeSize + i) * pieceSize;
-        const point = pathElement!.getPointAtLength(distance);
-        coords.push(point);
+      let current = 0;
+      for (let distance = 0; distance <= totalLength; distance += sampleRate) {
+        const sample = pathElement.getPointAtLength(distance);
+        const position = sample[segmentation];
+        while (current < numSegments - 1 && isBeyondSegments[current + 1](position)) {
+          current += 1;
+        }
+        lineSegments[current].push(sample);
       }
+    } else {
+      // segmentation === "length"
+      const numPointsInSegment = pointsInSegments.map(points => points.length);
+      const numPoints = numPointsInSegment.reduce((sum, curr) => sum + curr, 0);
+      const lengthBetweenPoints = totalLength / (numPoints - 1);
 
-      cumulativeSize += segmentPointCount;
+      const segmentBegins = numPointsInSegment.slice(0, numSegments - 1);
+      segmentBegins.unshift(0);
+      for (let i = 2; i < numSegments; i += 1) {
+        segmentBegins[i] += segmentBegins[i - 1];
+      }
+      for (let i = 0; i < numSegments; i += 1) {
+        segmentBegins[i] *= lengthBetweenPoints;
+      }
 
-      return coords;
-    });
+      let current = 0;
+      for (let distance = 0; distance <= totalLength; distance += sampleRate) {
+        const sample = pathElement.getPointAtLength(distance);
+        while (current < numSegments - 1 && distance >= segmentBegins[current + 1]) {
+          current += 1;
+        }
+        lineSegments[current].push(sample);
+      }
+    }
 
     return lineSegments;
   } catch (e) {
     return [];
   }
 }
-
-export default memoize(
```

Comment:
The rewritten `SplitLinePath` let us use `useMemo` there so no more need to `memoize` here.